### ISSUE TITLE
DataTable: Reflow Lost While Dragging Row

### ISSUE THREAD

**Issue by mossaab0** [primefaces/primefaces#7444]
**Describe the defect**
While dragging a row in a `p:dataTable` that has both `draggableRows` and `reflow` enabled, the reflow is lost for that particular row during the transition.
```
<p:dataTable draggableRows="true" reflow="true" [...] />
```

- The right side of the screenshot shows the row (highlighted in green) in the reflow mode.
- The left side shows the same row (highlighted in red) where the reflow is lost during dragging operation.

![dataTable draggableRows reflow](https://user-images.githubusercontent.com/2341730/122548035-c0fce300-d028-11eb-98a1-62b8fb2b3e9d.png)

**Environment:**
 - PF Version: 10.0.0
 - JSF Version: joinfaces 4.5.1 (jsf-spring-boot-starter, mojarra-spring-boot-starter).

**Comment by melloware** [comment]
I am not sure these datatable features were designed to work together.  Not all datatable features can be used together.

**Comment by mossaab0** [comment]
@melloware you mean this issue should be marked as a feature request?

**Comment by melloware** [comment]
Yes i changed it to New Feature as these are currently not designed to be used together.  I doubt anyone in the community will work on it so most likely you would have to sponsor that feature via PRO or submit a PR.

**Comment by melloware** [comment]
@mossaab0 OK this MonkeyPatch looks big but can you try it and see if it fixes your issue?  Just add this script to your application JS file.
 
```javascript
if (PrimeFaces.widget.DataTable) {
    PrimeFaces.widget.DataTable.prototype.makeRowsDraggable = function() {
        var $this = this,
            draggableHandle = this.cfg.rowDragSelector || 'td,span:not(.ui-c)';

        this.tbody.sortable({
            placeholder: 'ui-datatable-rowordering ui-state-active',
            cursor: 'move',
            handle: draggableHandle,
            appendTo: document.body,
            start: function(event, ui) {
                ui.helper.css('z-index', PrimeFaces.nextZindex());
            },
            helper: function(event, ui) {
                var cells = ui.children(),
                    helper = $('<div class="ui-datatable ui-widget"><table><tbody class="ui-datatable-data"></tbody></table></div>'),
                    helperRow = ui.clone(),
                    helperCells = helperRow.children();

                for (var i = 0; i < helperCells.length; i++) {
                    var helperCell = helperCells.eq(i);
                    helperCell.width(cells.eq(i).width());
                    // #5584 reflow must remove column title span
                    helperCell.children().remove('.ui-column-title');
                }

                helperRow.appendTo(helper.find('tbody'));

                return helper;
            },
            update: function(event, ui) {
                var fromIndex = ui.item.data('ri'),
                    fromNode = ui.item;
                itemIndex = ui.item.index(),
                    toIndex = $this.paginator ? $this.paginator.getFirst() + itemIndex : itemIndex;
                isDirectionUp = fromIndex >= toIndex;

                // #5296 must not count header group rows
                // #6557 must not count expanded rows
                if (isDirectionUp) {
                    for (i = 0; i <= toIndex; i++) {
                        fromNode = fromNode.next('tr');
                        if (fromNode.hasClass('ui-rowgroup-header') || fromNode.hasClass('ui-expanded-row-content')) {
                            toIndex--;
                        }
                    }
                } else {
                    fromNode.prevAll('tr').each(function() {
                        var node = $(this);
                        if (node.hasClass('ui-rowgroup-header') || node.hasClass('ui-expanded-row-content')) {
                            toIndex--;
                        }
                    });
                }
                toIndex = Math.max(toIndex, 0);

                $this.syncRowParity();

                var options = {
                    source: $this.id,
                    process: $this.id,
                    params: [{
                            name: $this.id + '_rowreorder',
                            value: true
                        },
                        {
                            name: $this.id + '_fromIndex',
                            value: fromIndex
                        },
                        {
                            name: $this.id + '_toIndex',
                            value: toIndex
                        },
                        {
                            name: this.id + '_skipChildren',
                            value: true
                        }
                    ]
                }

                if ($this.hasBehavior('rowReorder')) {
                    $this.callBehavior('rowReorder', options);
                } else {
                    PrimeFaces.ajax.Request.handle(options);
                }

                // #7444 must reflow after update
                $this.postUpdateData();
            },
            change: function(event, ui) {
                if ($this.cfg.scrollable) {
                    PrimeFaces.scrollInView($this.scrollBody, ui.placeholder);
                }
            }
        });
    }
}
```

**Comment by melloware** [comment]
Oh nevermind.   I just tested it its working fine and the dragged row is dropped correctly right?  You are just saying while dragging the row the reflow is lost correct?  But everything works fine when its dropped right?

**Comment by mossaab0** [comment]
@melloware while dragging, the same unwanted behavior happens (i.e., reflow is lost). After dropping, reflow is recovered. But, each column name gets duplicated for all rows. When I re-drag again, a third column name is added (for each column and row).

**Comment by mossaab0** [comment]
> @mossaab0 OK this MonkeyPatch looks big but can you try it and see if it fixes your issue? Just add this script to your application JS file.
> 
> ```js
> if (PrimeFaces.widget.DataTable) {
>     PrimeFaces.widget.DataTable.prototype.makeRowsDraggable = function() {
>         var $this = this,
>             draggableHandle = this.cfg.rowDragSelector || 'td,span:not(.ui-c)';
> 
>         this.tbody.sortable({
>             placeholder: 'ui-datatable-rowordering ui-state-active',
>             cursor: 'move',
>             handle: draggableHandle,
>             appendTo: document.body,
>             start: function(event, ui) {
>                 ui.helper.css('z-index', PrimeFaces.nextZindex());
>             },
>             helper: function(event, ui) {
>                 var cells = ui.children(),
>                     helper = $('<div class="ui-datatable ui-widget"><table><tbody class="ui-datatable-data"></tbody></table></div>'),
>                     helperRow = ui.clone(),
>                     helperCells = helperRow.children();
> 
>                 for (var i = 0; i < helperCells.length; i++) {
>                     var helperCell = helperCells.eq(i);
>                     helperCell.width(cells.eq(i).width());
>                     // #5584 reflow must remove column title span
>                     helperCell.children().remove('.ui-column-title');
>                 }
> 
>                 helperRow.appendTo(helper.find('tbody'));
> 
>                 return helper;
>             },
>             update: function(event, ui) {
>                 var fromIndex = ui.item.data('ri'),
>                     fromNode = ui.item;
>                 itemIndex = ui.item.index(),
>                     toIndex = $this.paginator ? $this.paginator.getFirst() + itemIndex : itemIndex;
>                 isDirectionUp = fromIndex >= toIndex;
> 
>                 // #5296 must not count header group rows
>                 // #6557 must not count expanded rows
>                 if (isDirectionUp) {
>                     for (i = 0; i <= toIndex; i++) {
>                         fromNode = fromNode.next('tr');
>                         if (fromNode.hasClass('ui-rowgroup-header') || fromNode.hasClass('ui-expanded-row-content')) {
>                             toIndex--;
>                         }
>                     }
>                 } else {
>                     fromNode.prevAll('tr').each(function() {
>                         var node = $(this);
>                         if (node.hasClass('ui-rowgroup-header') || node.hasClass('ui-expanded-row-content')) {
>                             toIndex--;
>                         }
>                     });
>                 }
>                 toIndex = Math.max(toIndex, 0);
> 
>                 $this.syncRowParity();
> 
>                 var options = {
>                     source: $this.id,
>                     process: $this.id,
>                     params: [{
>                             name: $this.id + '_rowreorder',
>                             value: true
>                         },
>                         {
>                             name: $this.id + '_fromIndex',
>                             value: fromIndex
>                         },
>                         {
>                             name: $this.id + '_toIndex',
>                             value: toIndex
>                         },
>                         {
>                             name: this.id + '_skipChildren',
>                             value: true
>                         }
>                     ]
>                 }
> 
>                 if ($this.hasBehavior('rowReorder')) {
>                     $this.callBehavior('rowReorder', options);
>                 } else {
>                     PrimeFaces.ajax.Request.handle(options);
>                 }
> 
>                 // #7444 must reflow after update
>                 $this.postUpdateData();
>             },
>             change: function(event, ui) {
>                 if ($this.cfg.scrollable) {
>                     PrimeFaces.scrollInView($this.scrollBody, ui.placeholder);
>                 }
>             }
>         });
>     }
> }
> ```

@melloware while dragging, the same unwanted behavior happens (i.e., reflow is lost). After dropping, reflow is recovered. But, each column name gets duplicated for all rows. When I re-drag again, a third column name is added (for each column and row).

**Comment by melloware** [comment]
OK my fix is not necessary. All you are saying is the table looks funny while dragging right?  But after you drop eveyrthing looks correct right?

That is because when doing drag and drop a fake Helper row is rendered with HTML like this..
```javascript
var cells = ui.children(),
                    helper = $('<div class="ui-datatable ui-widget"><table><tbody class="ui-datatable-data"></tbody></table></div>'),
                    helperRow = ui.clone(),
                    helperCells = helperRow.children();
```

I don't think this is worth trying to fix while dragging if in reflow mode when its only whle you are dragging the rendering is off.

**Comment by melloware** [comment]
I thought you were originally saying the renderng of the table was wrong AFTER you dropped the row.  Its only message up while dragging.

**Comment by mossaab0** [comment]
> OK my fix is not necessary. All you are saying is the table looks funny while dragging right? But after you drop eveyrthing looks correct right?
> 
> That is because when doing drag and drop a fake Helper row is rendered with HTML like this..
> 
> ```js
> var cells = ui.children(),
>                     helper = $('<div class="ui-datatable ui-widget"><table><tbody class="ui-datatable-data"></tbody></table></div>'),
>                     helperRow = ui.clone(),
>                     helperCells = helperRow.children();
> ```
> 
> I don't think this is worth trying to fix while dragging if in reflow mode when its only whle you are dragging the rendering is off.

Your understanding is correct. But I disagree with your thinking. When the user drags the row and scrolls up/down to drop it, the row gets overflown horizontal, and most of the content of the row isn't visible to them anymore. That makes it difficult for the user to decide where to drop the row (e.g., they can't compare with the content of the other rows).

I understand that you might consider this a rare use case, giving it a low priority. But I still believe it's worth fixing, and is not merely an aesthetic thing. In other words, the user does have a high cognitive load dragging a row this "funny" way (with a large number of columns, which is why `reflow` was needed in the first place).

**Comment by melloware** [comment]
Understood!  Well all the draggable code is there if you can figure it out we would gladly accept a PR!

**Comment by mossaab0** [comment]
@melloware I think I have a better understanding of what's happening, and I believe it'd be an easy clean fix for somebody who is more experienced than I am (I'm not a front end person).

When the row is dragged, this `div` is created as an immediate child of `body` and becomes what the user is dragging:

```
<div class="ui-datatable ui-widget ui-sortable-helper" style="width: 524.8px; height: 340.6px; position: absolute; z-index: 1003; left: 18.4px; top: 644.975px;">
```

Note that the children of the newly created div don't have any ancestor with `class=ui-datatable-reflow`. Hence, media queries such as the following are not triggered:

```
@media (max-width: 640px)
.ui-datatable-reflow .ui-datatable-data td[role="gridcell"]:not(.ui-helper-hidden) {
    text-align: left;
    display: block;
    border: 0 none;
    width: 100%;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    float: left;
    clear: left;
}
```

A partial workaround for me was to duplicate the CSS code of `.ui-datatable-reflow` and apply it to `.ui-sortable-helper`. While this has partially solved the problem (i.e., the dragged row has the reflow style, but the column names have disappeared), I believe a cleaner solution should be trivial for a more experienced developer.

That ideal solution would have two changes:
1. Make the `ui-sortable-helper` div a child of a `ui-datatable-reflow` div.
2. Include the `<span class="ui-column-title"/>` block inside the `ui-sortable-helper` div.

**Comment by mossaab0** [comment]
@melloware copying the above `@media (max-width: 640px) .ui-datatable-reflow .ui-datatable-data td[role="gridcell"]` code to `.ui-sortable-placeholder td` gave an extra visual improvement.

**Comment by melloware** [comment]
Can you paste your exact CSS changes here in complete.

**Comment by melloware** [comment]
Or provide a PR.

**Comment by mossaab0** [comment]
No. A mere CSS change would force the dragged row to appear in the `reflow` mode even when `reflow=false`.

Instead, the code that inserts `<div class="ui-sortable-helper"/>` as a child of `<body/>` should instead insert it as a child of  `<div class="ui-datatable-reflow"/>`.

I'm sorry. I don't know how to do it (for my use case, `reflow=true`, so I don't care about breaking the `reflow=false` case).

Feel free to close the ticket if you're not available to work on a clean fix.

**Comment by melloware** [comment]
It was a CSS fix and you had to modify this right? `helperRow.appendTo(helper.find('tbody'));`Just trying to get to the bottom of what you are fixing but its hard to follow with all your comments and suggestions exactly what you are touching.

**Comment by mossaab0** [comment]
I added the following to my CSS file. This forces the dragged row to appear in the `reflow` mode regardless of the value of `reflow` in `p:dataTable`.

```
@media (max-width: 640px) {
    .ui-sortable-helper .ui-datatable-data td[role="gridcell"],
    .ui-sortable-placeholder td {
    text-align: left;
    display: block;
    border: 0 none;
    width: 100%;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    float: left;
    clear: left;
    }

    .ui-sortable-helper .ui-datatable-data td[role="gridcell"] .ui-column-title {
        margin: -.4em 1em -.4em -.4em;
        color: gray;
        font-size: 0.8em;
        padding: .4em !important;
        min-width: 30% !important;
        display: inline-block !important;
    }
}
```

**Comment by melloware** [comment]
So I think we need two things..

1. Change this line of code if `cfg.reflow`...

```javascript
helper = $('<div class="ui-datatable ui-widget"><table><tbody class="ui-datatable-data"></tbody></table></div>'),
```
 to

```javascript
helper = $('<div class="ui-datatable ui-widget"><table><tbody class="ui-datatable-data ui-datatable-reflow"></tbody></table></div>'),
```

2. Change the CSS to have the `ui-datatable-reflow` style.

```css
@media (max-width: 640px) {
    .ui-sortable-helper .ui-datatable-data.ui-datatable-reflow td[role="gridcell"],
    .ui-sortable-placeholder td {
    text-align: left;
    display: block;
    border: 0 none;
    width: 100%;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    float: left;
    clear: left;
    }

    .ui-sortable-helper .ui-datatable-data.ui-datatable-reflow td[role="gridcell"] .ui-column-title {
        margin: -.4em 1em -.4em -.4em;
        color: gray;
        font-size: 0.8em;
        padding: .4em !important;
        min-width: 30% !important;
        display: inline-block !important;
    }
}
```

**Comment by mossaab0** [comment]
Before dragging, we currently have this structure:

```
<body>
    <div class="ui-datatable  ui-datatable-reflow">
        <div class="ui-datatable-header"/>
        <div class="ui-datatable-tablewrapper">
            <table>
                <thead/>
                <tbody class="ui-datatable-data"/>
                    <tr data-ri="0"/>
                    <tr data-ri="1"/>
                </tbody>
            </table>
        </div>
    </div>
</body>
```

While dragging, we currently get this structure:
```
<body>
    <div class="ui-datatable  ui-datatable-reflow">
        <div class="ui-datatable-header"/>
        <div class="ui-datatable-tablewrapper">
            <table>
                <thead/>
                <tbody class="ui-datatable-data"/>
                    <tr data-ri="0"/>
                    <tr data-ri="1"/>
                    **<tr class="ui-sortable-placeholder"/>**
                </tbody>
            </table>
        </div>
    </div>
    **<div class="ui-sortable-helper"/>**
</body>
```

Instead, I propose the following, which doesn't involve making any CSS change to `ui-sortable-placeholder`:

Solution 1:
```
<body>
    <div class="ui-datatable  ui-datatable-reflow">
        <div class="ui-datatable-header"/>
        <div class="ui-datatable-tablewrapper">
            <table>
                <thead/>
                <tbody class="ui-datatable-data"/>
                    <tr data-ri="0"/>
                    <tr data-ri="1"/>
                    **<tr class="ui-sortable-placeholder"/>**
                    **<div class="ui-sortable-helper"/>**
                </tbody>
            </table>
        </div>
    </div>
</body>
```

Solution 2:
```
<body>
    <div class="ui-datatable  ui-datatable-reflow">
        <div class="ui-datatable-header"/>
        <div class="ui-datatable-tablewrapper">
            <table>
                <thead/>
                <tbody class="ui-datatable-data"/>
                    <tr data-ri="0"/>
                    <tr data-ri="1"/>
                    **<tr class="ui-sortable-placeholder"/>**
                </tbody>
            </table>
        </div>
    </div>
    **<div class="ui-datatable-reflow"><div class="ui-sortable-helper"/></div>**
</body>
```

**Comment by melloware** [comment]
That DIV is created by the Jquery UI Sortable and its appended to body here.

```javascript
appendTo: document.body,
```

See: https://api.jqueryui.com/sortable/#option-appendTo

**Comment by mossaab0** [comment]
If I understand your comment correctly, you're perhaps suggesting we can do Solution 1, but not Solution 2. That's fine with me.

**Comment by mossaab0** [comment]
Is this Solution 3 possible?

```
<body>
    <div class="ui-datatable  ui-datatable-reflow">
        <div class="ui-datatable-header"/>
        <div class="ui-datatable-tablewrapper">
            <table>
                <thead/>
                <tbody class="ui-datatable-data"/>
                    <tr data-ri="0"/>
                    <tr data-ri="1"/>
                    **<tr class="ui-sortable-placeholder"/>**
                </tbody>
            </table>
        </div>
    </div>
    **<div class="ui-sortable-helper ui-datatable-reflow"/>**
</body>
```

**Comment by melloware** [comment]
I don't think Solution 3 is possible.

**Comment by mossaab0** [comment]
Well, if Solution 1 is possible, that should work, I hope!

**Comment by melloware** [comment]
well maybe 3 is possible.  Trying to determine that.

**Comment by melloware** [comment]
@mossaab0 I could never get this to work right.  IF interested in submitting a PR that would be great.  If not I am inclined to close this as wontfix.

**Comment by melloware** [comment]
Closing as wontfix for all the reasons listed in the ticket.

**Comment by BalusC** [comment]
I stumbled upon this as well, without monkeypatching I came to this work around:

```js
$(document).on("sortstart", ".ui-datatable-reflow .ui-sortable", function(event, ui) {
    PrimeFaces.queueTask(() => {
        $(".ui-sortable-helper").addClass("ui-datatable-reflow");
        $(".ui-sortable-placeholder").find("td").attr("role", "gridcell");
    });
});
```

```css
@media (max-width: 640px) {
    body .ui-datatable-reflow .ui-datatable-data td[role="gridcell"]:not(.ui-helper-hidden) {
        display: flex;
    }

    body .ui-datatable-reflow .ui-datatable-data .ui-sortable-placeholder td {
        padding: .5em;
        border: 0;
    }

    body .ui-datatable-reflow.ui-sortable-helper,
    body .ui-datatable-reflow.ui-sortable-helper td {
        max-width: 90vw;
    }
}
```

Not perfect, but much more acceptable than existing behavior.

**Comment by melloware** [comment]
@BalusC PR submitted based on your feedback can you review?

**Comment by baukescholtz** [comment]
Ah that could work .. but it's "not perfect". Especially:

1. column titles are missing in the sortable helper (the droppable item), they need to be copied from original columns
2. width is not correct per se, the default sortable helper has mis-calculated widths "hardcoded" via $element.width(...), I haven't analyzed why these widths are so off, I believe it miscalculated because the initial sortable helper was missing the ui-datatable-reflow class, ideally the ui-datatable-reflow class should be set during the creation of the sortable helper, before its widths are calculated

**Comment by BalusC** [comment]
Ah that could work .. but it's "not perfect". Especially:

1. column titles are missing in the sortable helper (the droppable item), they need to be copied from original columns
2. width is not correct per se, the default sortable helper has mis-calculated widths "hardcoded" via $element.width(...), I haven't analyzed why these widths are so off, I believe it miscalculated because the initial sortable helper was missing the ui-datatable-reflow class, ideally the ui-datatable-reflow class should be set during the creation of the sortable helper, before its widths are calculated
